core/src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -19,281 +19,590 @@
-                }
+            if (options.containsKey(CONTENT_TYPE_OPTION)) {
+                XContentType contentType = XContentType.fromMediaTypeOrFormat(options.get(CONTENT_TYPE_OPTION));
+                out.writeBoolean(true);
+                contentType.writeTo(out);
             } else {
-                throw new ElasticsearchParseException("unexpected field [{}]", cfn);
+                out.writeBoolean(false);
             }
         }
-        if (script == null) {
-            throw new ElasticsearchParseException("expected one of [{}], [{}] or [{}] fields, but found none",
-                    ScriptType.INLINE.getParseField() .getPreferredName(), ScriptType.FILE.getParseField().getPreferredName(),
-                    ScriptType.STORED.getParseField() .getPreferredName());
-        }
-        return new Script(script, type, lang, params, contentType);
     }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(lang, params, script, type, contentType);
+    /**
+     * @return The id for this {@link Script} if the {@link ScriptType} is {@link ScriptType#FILE} or {@link ScriptType#STORED}.
+     *         The code for this {@link Script} if the {@link ScriptType} is {@link ScriptType#INLINE}.
+     */
+    public String getIdOrCode() {
+        return idOrCode;
     }
 
-    @Override
-    public boolean equals(Object obj) {
-        if (this == obj) return true;
-        if (obj == null) return false;
-        if (getClass() != obj.getClass()) return false;
-        Script other = (Script) obj;
-
-        return Objects.equals(lang, other.lang) &&
-                Objects.equals(params, other.params) &&
-                Objects.equals(script, other.script) &&
-                Objects.equals(type, other.type) &&
-                Objects.equals(contentType, other.contentType);
+    /**
+     * @return The {@link ScriptType} for this {@link Script}.
+     */
+    public ScriptType getType() {
+        return type;
+    }
+
+    /**
+     * @return The language for this {@link Script}.
+     */
+    public String getLang() {
+        return lang;
+    }
+
+    /**
+     * @return The map of compiler options for this {@link Script}.
+     */
+    public Map<String, String> getOptions() {
+        return options;
+    }
+
+    /**
+     * @return The map of user-defined params for this {@link Script}.
+     */
+    public Map<String, Object> getParams() {
+        return params;
     }
 
     @Override
-    public String toString() {
-        return "[script: " + script + ", type: " + type.getParseField().getPreferredName() + ", lang: "
-                + lang + ", params: " + params + "]";
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        Script script = (Script)o;
+
+        if (type != script.type) return false;
+        if (!lang.equals(script.lang)) return false;
```

Comment:
I auto generate these through Intellij, so I'm going to leave this as is.  If this needs to be generated again we lose any customization.